### PR TITLE
HPCC-18090 Support get_session_timeout and ECLWatchAutoRefresh in esp

### PR DESCRIPTION
**Pull request by wangkx**
Also add code to check service name when checking logout method.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead please review.

File: esp/bindings/http/platform/httpservice.cpp [hunk 1]
```diff
@@ -1136,6 +1136,30 @@ EspAuthState CEspHttpServer::authNewSession(EspAuthRequest& authReq, const char*
     return authSucceeded;
 }
 
+void CEspHttpServer::createGetSessionTimeoutResponse(StringBuffer& resp, ESPSerializationFormat format, IPropertyTree* sessionTree)
+{
+    __int64 timeoutAt = sessionTree ? sessionTree->getPropInt64(PropSessionTimeoutAt, 0) : 0;
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: A general comment, I like to see more comments in the code.  In this instance, what is the meaning of 0 here?

File: esp/bindings/http/platform/httpservice.cpp [hunk 2]
```diff
@@ -1136,6 +1136,30 @@ EspAuthState CEspHttpServer::authNewSession(EspAuthRequest& authReq, const char*
...
+    bool timeoutByAdmin = sessionTree ? sessionTree->getPropBool(PropSessionTimeoutByAdmin, false) : false;
+    if (format == ESPSerializationJSON)
+    {
+        resp.set("{ ");
+        resp.append(" \"GetSessionTimeoutResponse\": { ");
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: appendf  is more efficient, and probably more readable

File: esp/bindings/http/platform/httpservice.cpp [hunk 3]
```diff
@@ -1136,6 +1136,30 @@ EspAuthState CEspHttpServer::authNewSession(EspAuthRequest& authReq, const char*
...
+        resp.append("\"TimeoutAt\": ").append(timeoutAt);
+        if (timeoutByAdmin)
+            resp.appendf(", \"TimeoutByAdmin\": true");
+        resp.append(" }");
+        resp.append(" }");
+    }
+    else
+    {
+        resp.set("<GetSessionTimeoutResponse>");
+        resp.append("<TimeoutAt>").append(timeoutAt).append("</TimeoutAt>");
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: appendf

File: esp/bindings/http/platform/httpservice.cpp [hunk 4]
```diff
@@ -1176,17 +1210,24 @@ EspAuthState CEspHttpServer::authExistingSession(EspAuthRequest& authReq, unsign
...
+        m_request->getParameter("ECLWatchAutoRefresh", autoRefresh);
+
         CDateTime now;
         now.setNow();
         time_t createTime = now.getSimple();
         sessionTree->setPropInt64(PropSessionLastAccessed, createTime);
-        if (!sessionTree->getPropBool(PropSessionTimeoutByAdmin, false))
-            sessionTree->setPropInt64(PropSessionTimeoutAt, createTime + authReq.authBinding->getSessionTimeoutSeconds());
-
+        if (!sessionTree->getPropBool(PropSessionTimeoutByAdmin, false) && (autoRefresh.isEmpty() || strieq(autoRefresh.str(), "0")))
```
Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: I am not clear on what autoRefresh can be set to.  Numeric strings?

**Review by RussWhitehead — CHANGES_REQUESTED** [review]

File: esp/bindings/http/platform/httpservice.cpp (hunk 1 above)
Comment (wangkx) [inline esp/bindings/http/platform/httpservice.cpp]: Based on my agreement with Miguel, I returns the time stamp for the timeoutAt and the 0 means that the session is timeout. I will add some comment here.

File: esp/bindings/http/platform/httpservice.cpp (hunk 4 above)
Comment (wangkx) [inline esp/bindings/http/platform/httpservice.cpp]: The authRefresh is flag: '1' means yes and '0' means no.

Comment (RussWhitehead) [inline esp/bindings/http/platform/httpservice.cpp]: Please add that as a comment

**Comment by wangkx** [comment]
@RussWhitehead please review my changes.

**Comment by RussWhitehead** [comment]
@wangkx Looks good

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

**Pull request merged**